microsoft/onefuzz
# Exponential backoff for failed notifications

## Summary of the Pull Request

_What is this about?_

#2553 

Today, our file-changes-poison queue does not get consumed and we manually re-queue those messages. This PR automatically re-processes those messages using an exponential backoff. The backoff times are:
```
// 1 - 5 mins
// 2 - 25 mins
// 3 - 2 hours 5 mins
// 4 - 10 hours 25 mins
// 5 - 2 days 4 hours 5 mins
// >6 - 7 days +/- 1 day of random variance so all the messages don't dequeue at the same time
```

Once a message hits the 1 week backoff point (or 6 attempts), we log a warning (though we should consider adding a webhook as well).

## Info on Pull Request

_What does this include?_

* Consuming the `file-changes-poison` queue
* Exponential backoff retry for poison queue messages
* Logging when poison queue messages are _really_ failing
* Rename `failTaskOnTransientError` to `isLastRetryAttempt` to more accurately reflect the behavior of that flag (it was changed in a previous PR)

## Validation Steps Performed

_How does someone test & validate?_

* Enqueue a poison queue message
* Run the C# code such that it will fail when processing the message (aka: always throw an exception)
* Validate the message was consumed from the queue
* Validate the message was put back on the queue but hidden (azure portal shows 0 of 1 messages)
* Re-run the C# code but don't throw any exceptions
* Validate the message was consumed from the queue after the visibility period and make sure it was created successfully and not put back in the poison queue
